From 66ef8446efa15ca3893c04710ea096d2be4a3736 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Wed, 4 Dec 2024 17:54:39 +0800
Subject: [PATCH 042/119] mtk: mt76: mt7996: add beacon waiting mechanism for
 active scanning on radar channel

For radar/no-IR channels, the scan defaults to passive scanning.
However, if beacons/probe resps are received on this channel, then
the scan can be upgraded to active scanning.
This helps the STA scan for the hidden root AP on DFS channels.

Clear MT76_SCANNING_WAIT_BEACON & MT76_SCANNING_BEACON_DONE flag if
scanning is completed

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mt76.h       |  2 ++
 mt7996/mac.c | 23 +++++++++++++++++++++++
 2 files changed, 25 insertions(+)

diff --git a/mt76.h b/mt76.h
index ff48ccea..c65e258e 100644
--- a/mt76.h
+++ b/mt76.h
@@ -565,6 +565,8 @@ enum {
 	MT76_STATE_ROC,
 	MT76_STATE_PM,
 	MT76_STATE_WED_RESET,
+	MT76_SCANNING_WAIT_BEACON,
+	MT76_SCANNING_BEACON_DONE,
 };
 
 enum mt76_sta_event {
diff --git a/mt7996/mac.c b/mt7996/mac.c
index ec1f2d36..291f38e6 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -13,6 +13,7 @@
 #include "mcu.h"
 #include "vendor.h"
 #include "mt7996_trace.h"
+#include "debug.h"
 
 static const struct mt7996_dfs_radar_spec etsi_radar_specs = {
 	.pulse_th = { 110, -10, -80, 40, 5200, 128, 5200 },
@@ -53,6 +54,8 @@ static const struct mt7996_dfs_radar_spec jp_radar_specs = {
 	},
 };
 
+static void mt7996_scan_rx(struct mt7996_phy *phy);
+
 static struct mt76_wcid *mt7996_rx_get_wcid(struct mt7996_dev *dev,
 					    u16 idx, u8 band_idx)
 {
@@ -645,6 +648,10 @@ mt7996_mac_fill_rx(struct mt7996_dev *dev, enum mt76_rxq_id q,
 
 		hdr = mt76_skb_get_hdr(skb);
 		fc = hdr->frame_control;
+
+		if (unlikely(ieee80211_is_probe_resp(fc) || ieee80211_is_beacon(fc)))
+			mt7996_scan_rx(phy);
+
 		if (ieee80211_is_data_qos(fc)) {
 			u8 *qos = ieee80211_get_qos_ctl(hdr);
 
@@ -3443,6 +3450,22 @@ void mt7996_mac_twt_teardown_flow(struct mt7996_dev *dev,
 	dev->twt.n_agrt--;
 }
 
+static void mt7996_scan_rx(struct mt7996_phy *phy)
+{
+	struct ieee80211_vif *vif = phy->scan_vif;
+	struct mt7996_vif *mvif;
+
+	if (!vif || !test_bit(MT76_SCANNING, &phy->mt76->state))
+		return;
+
+	if (test_and_clear_bit(MT76_SCANNING_WAIT_BEACON, &phy->mt76->state)) {
+		mvif = (struct mt7996_vif *)vif->drv_priv;
+		set_bit(MT76_SCANNING_BEACON_DONE, &phy->mt76->state);
+		cancel_delayed_work(&phy->scan_work);
+		ieee80211_queue_delayed_work(phy->mt76->hw, &phy->scan_work, 0);
+	}
+}
+
 static int
 mt7996_beacon_mon_send_probe(struct mt7996_phy *phy, struct mt7996_vif *mvif,
 			     struct ieee80211_bss_conf *conf, unsigned int link_id)
-- 
2.45.2

